### PR TITLE
integrate docker compose

### LINKED ISSUE #24469: Integrate Spring Boot's Docker Compose support
##### **Overview of the issue**

Add Docker Compose support so dependent containers are started automatically. At the bottom of [the docs](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#features.docker-compose), it mentions how you can point to a file. It'd be cool if multiple files are allowed.

##### **Motivation for or Use Case**

This is a better developer experience because you do t have to start anything manually. It might help simplify our CI jobs too.

##### **JHipster Version(s)**

8.0.0

### PR DESCRIPTION
**Pull request by mshima**
Fixes #24469.

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by mraible** [comment]
Added bug bounty to the issue this fixes.

**Comment by mraible** [comment]
Is there anything I can do to help move this along?

**Comment by mshima** [comment]
@mraible its not clear to me how to test it.
I am thinking in adding tests to daily builds.

**Comment by mshima** [comment]
I know PostgreSQL is broken due to missing password.

**Comment by mraible** [comment]
We could remove the step to start Docker containers before the app starts. Maybe even remove the scripts to start containers since it'll be automatic.

**Comment by mshima** [comment]
We test compiled jar for monoliths and docker images for microservices.
The compiled jar and docker image removes docker-compose lib. So it work have docker compose integration.

I think this behavior is correct because the production jar and docker images are the most important.

To test docker compose integration we should run using `./mvnw`.

File: generators/spring-data-relational/support/database-data.ts [hunk 1]
```diff
@@ -151,6 +153,7 @@ const databaseData: Record<string, DatabaseDataSpec> = {
     jdbcDriver: 'org.postgresql.Driver',
     hibernateDialect: 'org.hibernate.dialect.PostgreSQLDialect',
     port: ':5432/',
+    defaultPassword: 'postgresql',
```
Comment (mshima) [inline generators/spring-data-relational/support/database-data.ts:156]: password is [required for databases](https://github.com/spring-projects/spring-boot/blob/ce01dbbad41496a2de6fb1c35bd08edfb8a8a430/spring-boot-project/spring-boot-docker-compose/src/test/java/org/springframework/boot/docker/compose/service/connection/mariadb/MariaDbEnvironmentTests.java#L59-L63).

**Comment by mshima** [comment]
I don't get why the application does not fails in some containers failures:
```
[0] Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled. 
[0] 2024-09-12T13:21:05.843Z ERROR 3295 --- [  restartedMain] o.s.boot.SpringApplication               : Application run failed 
[0] 
[0] org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in class path resource [org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaConfiguration.class]: Unable to create requested service [org.hibernate.engine.jdbc.env.spi.JdbcEnvironment] due to: Unable to determine Dialect without JDBC metadata (please set 'jakarta.persistence.jdbc.url' for common cases or 'hibernate.dialect' when a custom Dialect implementation must be provided)
[0] 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1806)
[0] 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:600)
[0] 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)
[0] 	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337)
[0] 	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
[0] 	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335)
[0] 	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:205)
[0] 	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:954)
[0] 	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:625)
[0] 	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146)
[0] 	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754)
[0] 	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456)
[0] 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:335)
[0] 	at com.mycompany.myapp.AppApp.main(AppApp.java:70)
[0] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[0] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[0] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[0] 	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
[0] 	at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:50)
[0] Caused by: org.hibernate.service.spi.ServiceException: Unable to create requested service [org.hibernate.engine.jdbc.env.spi.JdbcEnvironment] due to: Unable to determine Dialect without JDBC metadata (please set 'jakarta.persistence.jdbc.url' for common cases or 'hibernate.dialect' when a custom Dialect implementation must be provided)
[0] 	at org.hibernate.service.internal.AbstractServiceRegistryImpl.createService(AbstractServiceRegistryImpl.java:276)
[0] 	at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:238)
[0] 	at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:215)
[0] 	at org.hibernate.boot.model.relational.Database.<init>(Database.java:45)
[0] 	at org.hibernate.boot.internal.InFlightMetadataCollectorImpl.getDatabase(InFlightMetadataCollectorImpl.java:221)
[0] 	at org.hibernate.boot.internal.InFlightMetadataCollectorImpl.<init>(InFlightMetadataCollectorImpl.java:189)
[0] 	at org.hibernate.boot.model.process.spi.MetadataBuildingProcess.complete(MetadataBuildingProcess.java:171)
[0] 	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.metadata(EntityManagerFactoryBuilderImpl.java:1431)
[0] 	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.build(EntityManagerFactoryBuilderImpl.java:1502)
[0] 	at org.springframework.orm.jpa.vendor.SpringHibernateJpaPersistenceProvider.createContainerEntityManagerFactory(SpringHibernateJpaPersistenceProvider.java:75)
[0] 	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:390)
[0] 	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:409)
[0] 	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.afterPropertiesSet(AbstractEntityManagerFactoryBean.java:396)
[0] 	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.afterPropertiesSet(LocalContainerEntityManagerFactoryBean.java:366)
[0] 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1853)
[0] 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1802)
[0] 	... 18 common frames omitted
[0] Caused by: org.hibernate.HibernateException: Unable to determine Dialect without JDBC metadata (please set 'jakarta.persistence.jdbc.url' for common cases or 'hibernate.dialect' when a custom Dialect implementation must be provided)
[0] 	at org.hibernate.engine.jdbc.dialect.internal.DialectFactoryImpl.determineDialect(DialectFactoryImpl.java:191)
[0] 	at org.hibernate.engine.jdbc.dialect.internal.DialectFactoryImpl.buildDialect(DialectFactoryImpl.java:87)
[0] 	at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.getJdbcEnvironmentWithDefaults(JdbcEnvironmentInitiator.java:152)
[0] 	at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.getJdbcEnvironmentUsingJdbcMetadata(JdbcEnvironmentInitiator.java:362)
[0] 	at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:123)
[0] 	at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:77)
[0] 	at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.initiateService(StandardServiceRegistryImpl.java:130)
[0] 	at org.hibernate.service.internal.AbstractServiceRegistryImpl.createService(AbstractServiceRegistryImpl.java:263)
[0] 	... 33 common frames omitted
[0] 
[0] [INFO] ------------------------------------------------------------------------
[0] [INFO] BUILD SUCCESS
[0] [INFO] ------------------------------------------------------------------------
[0] [INFO] Total time:  01:17 min
[0] [INFO] Finished at: 2024-09-12T13:21:05Z
[0] [INFO] ------------------------------------------------------------------------
```

File: generators/spring-data-relational/support/database-data.ts (hunk 1 above)
Comment (mshima) [reply generators/spring-data-relational/support/database-data.ts]: Actually is required for postgresql only.
Fixed in v3.4 branch https://github.com/spring-projects/spring-boot/pull/41511?notification_referrer_id=NT_kwDOABWf37MxMTUyMzQ1NDEzMDoxNDE3MTgz#issuecomment-2346590381

**Comment by mshima** [comment]
- container integration workflow is conditional to `pr: docker-compose integration` label
- container integration workflow can have false positives
- we need to add a strong password to PostgreSQL to avoid sonar vulnerability and avoid Spring Boot issue
- MariaDB + reactive uses mysql driver, which doesn't match the url generated by the Services Connection Detail
- Cassandra is failing due to unknown issue
- if and serviceDiscovery is provided the integration fails
- keycloak is not supported

I plan to create a followup PR with improvements.

**Comment by mraible** [comment]
Is there an issue detailing why Keycloak doesn't work?

**Comment by mshima** [comment]
> Is there an issue detailing why Keycloak doesn't work?

https://github.com/spring-projects/spring-boot/issues/36777

File: generators/spring-data-relational/support/database-data.ts [hunk 2]
```diff
@@ -151,6 +151,9 @@ const databaseData: Record<string, DatabaseDataSpec> = {
     jdbcDriver: 'org.postgresql.Driver',
     hibernateDialect: 'org.hibernate.dialect.PostgreSQLDialect',
     port: ':5432/',
+    // Password is required by Spring Boot v3.3.x, can be removed for v3.4.x, see https://github.com/spring-projects/spring-boot/pull/41511
+    // Use a strong password to avoid being flagged by SonarQube
+    defaultPassword: 'yourStrong(!)Password',
```
Comment (mshima) [inline generators/spring-data-relational/support/database-data.ts:156]: This password is reported as sonar security issue.

**Comment by mraible** [comment]
When I first heard about this feature, I remember seeing it was possible to point at our existing Docker Compose files, which don't use testcontainers. Can we do that instead?

**Comment by mshima** [comment]
> When I first heard about this feature, I remember seeing it was possible to point at our existing Docker Compose files, which don't use testcontainers. Can we do that instead?

It's not using testcontainers at all, so I've recently renamed the sample/workflow to docker-compose-integration.
It's using the docker-compose file.
https://github.com/mshima/generator-jhipster/blob/0ac0cf69cb35872aebb85a6e1b5a8c7d0ec209f0/generators/spring-boot/templates/src/main/resources/config/application.yml.ejs#L281.

For each docker-compose container/service there is a Spring Boot service connection that detects required data like database/user/password/port (can be random) and configure the application according to it.
I couldn't find a way to add a generic container just to start and stop without a service connection.

**Comment by mraible** [comment]
The [issue you linked](https://github.com/spring-projects/spring-boot/issues/36777) to as Testcontainers in its description, that's what threw me off. The import and everything should work if it's just starting the existing Keycloak container. I can take a look.

**Comment by mraible** [comment]
When I create a new app (using `jhipster jdl blog-oauth2`), it shows the package.json has changed in red. It looks like a warning when it probably doesn't matter.

Screenshot 2024-09-12 at 20 15 04

**Comment by mraible** [comment]
It still has an eslint warning too:

```
npm warn ERESOLVE overriding peer dependency
npm warn While resolving: eslint-plugin-import@2.29.1
npm warn Found: eslint@9.9.1
npm warn node_modules/generator-jhipster/node_modules/eslint
npm warn   eslint@"9.9.1" from generator-jhipster@8.7.0
npm warn   node_modules/generator-jhipster
npm warn     dev generator-jhipster@"8.7.0" from the root project
npm warn
npm warn Could not resolve dependency:
npm warn peer eslint@"^2 || ^3 || ^4 || ^5 || ^6 || ^7.2.0 || ^8" from eslint-plugin-import@2.29.1
npm warn node_modules/generator-jhipster/node_modules/eslint-plugin-import
npm warn   eslint-plugin-import@"2.29.1" from generator-jhipster@8.7.0
npm warn   node_modules/generator-jhipster
npm warn
npm warn Conflicting peer dependency: eslint@8.57.0
npm warn node_modules/eslint
npm warn   peer eslint@"^2 || ^3 || ^4 || ^5 || ^6 || ^7.2.0 || ^8" from eslint-plugin-import@2.29.1
npm warn   node_modules/generator-jhipster/node_modules/eslint-plugin-import
npm warn     eslint-plugin-import@"2.29.1" from generator-jhipster@8.7.0
npm warn     node_modules/generator-jhipster
npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
```

**Comment by mraible** [comment]
I had to enable docker in `application-dev.yml`:

```yaml
spring:
  docker:
    compose:
      enabled: true
```

Lots of logging happens at the beginning, probably unrelated to this branch.

Screenshot 2024-09-12 at 20 24 52

Then, it hangs here. It might be trying to download Docker containers. If it is, I'd expect more logging.

Screenshot 2024-09-12 at 20 25 00

**Comment by mraible** [comment]
It works if I specify an absolute path to the file. It doesn't seem to work with a relative path. In `application-dev.yml`:

```yaml
spring:
  ...
  docker:
    compose:
      enabled: true
      file: /Users/mraible/Downloads/blog/src/main/docker/services.yml
```

**Comment by mshima** [comment]
> It still has an eslint warning too:
> 
> ```
> npm warn ERESOLVE overriding peer dependency
> npm warn While resolving: eslint-plugin-import@2.29.1
> npm warn Found: eslint@9.9.1
> npm warn node_modules/generator-jhipster/node_modules/eslint
> npm warn   eslint@"9.9.1" from generator-jhipster@8.7.0
> npm warn   node_modules/generator-jhipster
> npm warn     dev generator-jhipster@"8.7.0" from the root project
> npm warn
> npm warn Could not resolve dependency:
> npm warn peer eslint@"^2 || ^3 || ^4 || ^5 || ^6 || ^7.2.0 || ^8" from eslint-plugin-import@2.29.1
> npm warn node_modules/generator-jhipster/node_modules/eslint-plugin-import
> npm warn   eslint-plugin-import@"2.29.1" from generator-jhipster@8.7.0
> npm warn   node_modules/generator-jhipster
> npm warn
> npm warn Conflicting peer dependency: eslint@8.57.0
> npm warn node_modules/eslint
> npm warn   peer eslint@"^2 || ^3 || ^4 || ^5 || ^6 || ^7.2.0 || ^8" from eslint-plugin-import@2.29.1
> npm warn   node_modules/generator-jhipster/node_modules/eslint-plugin-import
> npm warn     eslint-plugin-import@"2.29.1" from generator-jhipster@8.7.0
> npm warn     node_modules/generator-jhipster
> npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
> ```

v8.7.0 release is installed.

**Comment by mraible** [comment]
This works:

```yaml
  docker:
    compose:
      enabled: true
      file: src/main/docker/services.yml
```

**Comment by mshima** [comment]
> Lots of logging happens at the beginning, probably unrelated to this branch.

Fixed in https://github.com/jhipster/generator-jhipster/pull/27267

**Comment by mraible** [comment]
> v8.7.0 release is installed.

I don't think so. I'm pulling from your `compose` branch, running `npm ci` and `npm link` before I run `jhipster`. I tried just now with `npm uninstall -g generator-jhipster` first to prove it's a local install and it has the same behavior.

**Comment by mraible** [comment]
I added the following to `logback-spring.xml` to see the downloading images progress.

```xml
<logger name="org.springframework.boot.docker" level="INFO"/>
```

Screenshot 2024-09-12 at 20 54 24

**Comment by mraible** [comment]
This might help: https://github.com/mshima/generator-jhipster/pull/1010

**Comment by mshima** [comment]
> This might help: [mshima#1010](https://github.com/mshima/generator-jhipster/pull/1010)

@mraible that config is already provided in application.yml

**Comment by mshima** [comment]
> > v8.7.0 release is installed.
> 
> I don't think so. I'm pulling from your `compose` branch, running `npm ci` and `npm link` before I run `jhipster`. I tried just now with `npm uninstall -g generator-jhipster` first to prove it's a local install and it has the same behavior.

Unless --skip-jhipster-dependencies is used, latest release will be installed.

**Comment by mraible** [comment]
> that config is already provided in application.yml

Shouldn't it only be in dev since it's only needed when using the dev profile?

**Comment by mraible** [comment]
> Unless --skip-jhipster-dependencies is used, latest release will be installed.

That's right. Sorry, I forgot about this after not working on the project for the last month. If I use this argument, the app is generated w/o the eslint warnings and starts with `./mvnw` as expected. It also fails with `mvn -Pprod`, which is expected. I also tried starting the containers with `docker compose -f src/main/docker/services.yml up` and starting with `mvn`. This works too, proving it should be backward compatible.

**Comment by mshima** [comment]
> > that config is already provided in application.yml
> 
> Shouldn't it be in application-dev.yml since it's only needed when using the dev profile?

It should work using prod profile for maven passing:
`./mvnw -Pprod,docker-compose`

For gradle it should work without docker-compose profile.

`docker-compose` dependency is in a profile due to https://github.com/GoogleContainerTools/jib-extensions/issues/158.

**Comment by mraible** [comment]
I tried to see if it's working with microservices. It appears there might be issues.

I generated microservices using `jhipster jdl reactive-mf --skip-jhipster-dependencies`. Then, I navigated into the gateway directory and ran `./gradlew`. I get an error:

```
java.lang.IllegalStateException: No Docker Compose file found in directory '/Users/mraible/Downloads/microfrontends/gateway/.'
```

The settings appear to be the same, so not sure why this is happening.

```
$ cat src/main/resources/config/application.yml | grep docker
  docker:
      file: src/main/docker/services.yml
```

Maybe this has something to do with setting the working directory in Gradle? https://github.com/springdoc/springdoc-openapi-gradle-plugin/issues/150

I don't see the problem if I re-generate the `blog-oauth2` monolith with Gradle.

**Comment by mshima** [comment]
@mraible see 2 latest commits for service discovery.
But seems to have a dependency problem now.

File: generators/common/__snapshots__/generator.spec.ts.snap [hunk 3]
```diff
@@ -154,6 +154,10 @@ sonar.exclusions = src/main/webapp/content/**/*.*, src/main/webapp/i18n/*.js, \\
 sonar.issue.ignore.multicriteria = \\
   S1192,S125,S3437,S4502,S4684,S5145,UndocumentedApi
```
Comment (DanielFran) [inline generators/common/__snapshots__/generator.spec.ts.snap:155]:
```suggestion
  S1192,S125,S3437,S4502,S4684,S5145,S6437,UndocumentedApi
```

**Review by DanielFran — CHANGES_REQUESTED** [review]

**Comment by mraible** [comment]
I tried generating microservices again with the latest code and the gateway seems to hang on startup.

Screenshot 2024-09-13 at 09 43 48

It eventually starts, but there's no logging about downloading containers. 

Screenshot 2024-09-13 at 09 45 02

If I try to start the "blog" microservice, it fails with the following error:

```
 Container blog-neo4j-1  Started
 Container blog-consul-config-loader-1  Started
Error response from daemon: driver failed programming external connectivity on endpoint blog-consul-1 (df4e3264b57f8306e3a92cc7ad5701ceb8ce9bdda8a1fcdb0a80216371b3c2c9): Bind for 127.0.0.1:8300 failed: port is already allocated

        at org.springframework.boot.docker.compose.core.ProcessRunner.run(ProcessRunner.java:96)
        at org.springframework.boot.docker.compose.core.DockerCli.run(DockerCli.java:80)
```

If I start the "store" microservice, the error is similar:

```
 Container store-consul-1  Starting
 Container store-consul-config-loader-1  Started
 Container store-mongodb-1  Started
Error response from daemon: driver failed programming external connectivity on endpoint store-keycloak-1 (0a8311dbce4c3f30e4ce5bbcd93400a17ce70c74ec7c6bec77efb2b0385b95e9): Bind for 127.0.0.1:9080 failed: port is already allocated

        at org.springframework.boot.docker.compose.core.ProcessRunner.run(ProcessRunner.java:96)
        at org.springframework.boot.docker.compose.core.DockerCli.run(DockerCli.java:80)
```

`docker ps` shows the following:

```
CONTAINER ID   IMAGE                                  COMMAND                  CREATED         STATUS                   PORTS                                                                                                        NAMES
42f3b4cec5ca   jhipster/consul-config-loader:v0.4.1   "/bin/sh -c /load-co…"   4 minutes ago   Up 4 minutes                                                                                                                          blog-consul-config-loader-1
c0d9b8482ed1   neo4j:5.23.0                           "tini -g -- /startup…"   4 minutes ago   Up 4 minutes (healthy)   127.0.0.1:7474->7474/tcp, 7473/tcp, 127.0.0.1:7687->7687/tcp                                                 blog-neo4j-1
66d3ccc65525   jhipster/consul-config-loader:v0.4.1   "/bin/sh -c /load-co…"   4 minutes ago   Up 4 minutes                                                                                                                          store-consul-config-loader-1
b8e596c9ea9f   mongo:7.0.6                            "docker-entrypoint.s…"   4 minutes ago   Up 4 minutes (healthy)   127.0.0.1:27017->27017/tcp                                                                                   store-mongodb-1
3f252b934acc   quay.io/keycloak/keycloak:25.0.1       "/opt/keycloak/bin/k…"   8 minutes ago   Up 8 minutes (healthy)   8080/tcp, 8443/tcp, 127.0.0.1:9080->9080/tcp, 9000/tcp, 127.0.0.1:9443->9443/tcp                             gateway-keycloak-1
e61812c3c2b3   postgres:16.4                          "docker-entrypoint.s…"   8 minutes ago   Up 8 minutes (healthy)   127.0.0.1:5432->5432/tcp                                                                                     gateway-postgresql-1
cffdb1ad1c97   bitnami/consul:1.19.2                  "/opt/bitnami/script…"   8 minutes ago   Up 8 minutes             127.0.0.1:8300->8300/tcp, 8301/udp, 127.0.0.1:8500->8500/tcp, 127.0.0.1:8600->8600/tcp, 8301/tcp, 8600/udp   gateway-consul-1
c4eb3697180d   jhipster/consul-config-loader:v0.4.1   "/bin/sh -c /load-co…"   8 minutes ago   Up 8 minutes                                                                                                                          gateway-consul-config-loader-1
```

I suspected this might happen, that's why I QA'd it. I believe we might have to remove some Docker Compose files from microservices because they'll already be started by the gateway.

File: generators/common/templates/sonar-project.properties.ejs [hunk 4]
```diff
@@ -55,19 +55,24 @@ sonar.issue.ignore.multicriteria=\
 <%_ } _%>
   S4684,S5145,UndocumentedApi
```
Comment (DanielFran) [inline generators/common/templates/sonar-project.properties.ejs:56]:
```suggestion
  S4684,S5145, \
<%_ if (backendTypeSpringBoot) { _%>
  S6437,\
<%_ } _%>
  UndocumentedApi
```

**Review by DanielFran — CHANGES_REQUESTED** [review]

**Comment by mshima** [comment]
> It eventually starts, but there's no logging about downloading containers.

I’ve reduced the log to warn, it’s too verbose in CI.

**Review by DanielFran — APPROVED** [review]

**Comment by mshima** [comment]
> I suspected this might happen, that's why I QA'd it. I believe we might have to remove some Docker Compose files from microservices because they'll already be started by the gateway.

microservices now depends on the database compose file instead of services.

**Review by DanielFran — APPROVED** [review]

**Comment by mraible** [comment]
I tested again with microservices and confirmed everything starts and works as expected. Nice work, @mshima!

**Review by mraible — APPROVED** [review]

**Pull request merged**